### DISCUSSION TITLE
Accessing compiled typedefs

### DISCUSSION THREAD

**Discussion by MerchantPaul** [CESNET/libyang#2039] [Q&A]
Is there a means to access a compiled typedef to retrieve the associated string from a numerical value?  For example, the **_chassis-id-subtype-type_**_ found in **ieee802-types@2022-10-31** contains values like [1] chassis-component, [2] interface-alias, etc,  I would like to know if I can retrieve string values from the numeric equivalent 1, 2, 3, ...   Also, if I have a string from this enumeration, say "mac-address", can I retrieve associated numerical value [4]?  What APIs, if any would allow me to accomplish this?

**Comment by michalvasko** [comment]
> Is there a means to access a compiled typedef...

What do you mean by compiled typedef? *libyang* first **parses** the typedef and then **compiles** any leaves/leaf-lists that use this typedef into their type.

In any case, to only work with the parsed typedef `struct lysp_tpdf` you first have to find it in a parsed module `struct lysp_module`, which is in `struct lys_module`. Then, get the `type` `struct lysp_type` from the typedef and there iterate over `enums` `struct lysp_type_enum`. In each there is `name` (the string name) and `value` (numeric value) for all the defined enums. So there are no functions for tasks like this, you will have to access the structures yourself.

If, however, you have an actual data node `struct lyd_node` that is using the typedef, its value is stored as `struct lysc_type_bitenum_item`, which includes the information about both name and value directly.
